src/command.cpp

```diff
@@ -134,24 +127,27 @@ shared_ptr<const CommandTable> Commands::get() {
 }
 
 void Commands::complete(Completion& completion) {
-    // wrap around complete_against_commands
-    // Once we have migrated all completions, we can implement command
-    // completion directly with the minimal interface that Completion provides.
-    // Then we can also unfriend this function from the Completion class.
-    char** argv = new char*[completion.args_.size() + 1];
-    argv[completion.args_.size()] = nullptr;
-    for (size_t i = 0; i < completion.args_.size(); i++) {
-        argv[i] = const_cast<char*>((completion.args_.begin() + i)->c_str());
-    }
-    int res = complete_against_commands(completion.args_.size(),
-                                        argv,
-                                        completion.index_,
-                                        completion.output_);
-    delete[] argv;
-    if (res == HERBST_NO_PARAMETER_EXPECTED) {
-        completion.none();
-    } else if (res != 0) {
-        completion.invalidArguments();
+    auto commandTable = Commands::get();
+    if (completion == 0) {
+        for (const auto& cmd : *commandTable) {
+            completion.full(cmd.first);
+        }
+    } else {
+        auto it = commandTable->find(completion[0]);
+        if (it == commandTable->end()) {
+            completion.invalidArguments();
+        } else if (it->second.hasCompletion()) {
+            // get new completion context with command name omitted.
+            Completion shifted = completion.shifted(1);
+            it->second.complete(shifted);
+            // TODO: call mergeResultsFrom(), once #1274 is merged.
```

Comment:
Now that #1274  is merged, do you want to update this PR or fix the TODO in a separate one?

Reply:
[x] done :-)

Reply:
Thanks :)